Add EIP: Builder Execution Requests (#11760)
ethereum/EIPs: EIPS/eip-8282.md

## Backwards Compatibility

This EIP is additive at the execution layer: it introduces new contracts at previously empty addresses. It does not modify the validator deposit contract at `0x00000000219ab540356cbb839cbe05303d7705fa`, the validator withdrawal/consolidation predeploys, or any existing validator's lifecycle.

At the consensus layer it modifies EIP-7732 (see [Changes to EIP-7732](#changes-to-eip-7732)): post-fork builder onboarding moves from the validator deposit request to `BUILDER_DEPOSIT_REQUEST_TYPE`, and builder exits move from the voluntary-exit operation to `BUILDER_EXIT_REQUEST_TYPE`. The fork-transition onboarding of builder-credentialed pending deposits is unchanged, so builders present at the fork are unaffected. The new request types are additive — blocks that contain no builder requests produce empty `request_data` for these types, which [EIP-7685](./eip-7685.md) excludes from the `requests_hash`.

> <!-- TODO: Add test cases and reference implementation -->

Reviewer: ```suggestion
## Reference Implementation

<!-- TODO: Add test cases and reference implementation -->
```

Adding this tag should fix the CI, allowing this EIP to be merged @wemeetagain 😄 👍

### Constants

The `0x03`/`0x04` request types MUST be unique across **all** active [EIP-7685](./eip-7685.md) request types, including in-flight proposals ([EIP-7804](./eip-7804.md), a Draft, also claims `0x03`); final allocation is coordinated in consensus-specs.

| Name | Value | Comment |
| --- | --- | --- |
| `BUILDER_DEPOSIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007732` <!-- placeholder --> | Predeploy address of the builder deposit contract |
| `BUILDER_EXIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007733` <!-- placeholder --> | Predeploy address of the builder exit contract |
| `BUILDER_DEPOSIT_REQUEST_TYPE` | `0x03`  <!-- placeholder --> | [EIP-7685](./eip-7685.md) request-type byte for builder deposits |
| `BUILDER_EXIT_REQUEST_TYPE` | `0x04` <!-- placeholder --> | [EIP-7685](./eip-7685.md) request-type byte for builder exits  |
| `SYSTEM_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe` | Address that invokes the end-of-block system call (as in [EIP-7002](./eip-7002.md)) |
| `MAX_DEPOSIT_REQUESTS_PER_BLOCK` | `256` | Maximum records the builder deposit contract drains into one block |

> | `TARGET_DEPOSIT_REQUESTS_PER_BLOCK` | `32` | Per-block request count above which the fee rises for the deposit contract |

Reviewer 2: - per https://github.com/ethereum/sys-asm/pull/50
```suggestion
| `MAX_DEPOSIT_REQUESTS_PER_BLOCK` | `64` | Maximum records the builder deposit contract drains into one block |
| `TARGET_DEPOSIT_REQUESTS_PER_BLOCK` | `8` | Per-block request count above which the fee rises for the deposit contract |
```

---
eip: 8282
title: Builder Execution Requests
description: Predeploy builder deposit and exit request contracts for EIP-7732 builders on the EIP-7685 request bus
author: Cayman (@wemeetagain), Nico Flaig <nflaig@protonmail.com>, Justin Traglia <jtraglia@ethereum.org>
discussions-to: https://ethereum-magicians.org/t/eip-8282-builder-execution-requests/28699
status: Draft
type: Standards Track
category: Core
created: 2026-05-22

> requires: 7685, 7732

Reviewer 3: ```suggestion
requires: 1559, 7685, 7732
```

Because you're directly referencing its algorithm.

Builders that must exist at the fork are unaffected: EIP-7732's fork-transition onboarding of builder-credentialed pending deposits is retained (see [Changes to EIP-7732](#changes-to-eip-7732)); only post-fork onboarding moves to the deposit contract. The deployed validator deposit contract is left untouched, and builder stake withdrawals continue to flow through EIP-7732's existing full-balance sweep.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).

### Constants

> All address and request-type values below are placeholders. The `0x03`/`0x04` request types MUST be unique across **all** active [EIP-7685](./eip-7685.md) request types, including in-flight proposals ([EIP-7804](./eip-7804.md), a Draft, also claims `0x03`); final allocation is coordinated in consensus-specs.

Reviewer 3: ```suggestion
The `0x03`/`0x04` request types MUST be unique across **all** active [EIP-7685](./eip-7685.md) request types, including in-flight proposals ([EIP-7804](./eip-7804.md), a Draft, also claims `0x03`); final allocation is coordinated in consensus-specs.
```

Proposals should be written as if they were already Final. It's too easy to miss updating one of these notes, and they look really unprofessional in Final EIPs.

Instead, you can use HTML-style comments (`<!-- TODO: replace placeholder -->`) and the linter will make sure you remove them before moving into Review.

> All address and request-type values below are placeholders. The `0x03`/`0x04` request types MUST be unique across **all** active [EIP-7685](./eip-7685.md) request types, including in-flight proposals ([EIP-7804](./eip-7804.md), a Draft, also claims `0x03`); final allocation is coordinated in consensus-specs.

Reviewer 3: > [...] a Draft [...]

Best not to mention the status of other proposals here. It's very likely that its status will change, making this sentence incorrect.

| Name | Value | Comment |
| --- | --- | --- |
| `BUILDER_DEPOSIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007732` | Predeploy address of the builder deposit contract (placeholder) |
| `BUILDER_EXIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007733` | Predeploy address of the builder exit contract (placeholder) |
| `BUILDER_DEPOSIT_REQUEST_TYPE` | `0x03` | [EIP-7685](./eip-7685.md) request-type byte for builder deposits (placeholder) |

> | `BUILDER_EXIT_REQUEST_TYPE` | `0x04` | [EIP-7685](./eip-7685.md) request-type byte for builder exits (placeholder) |

Reviewer 3: ```suggestion
| `BUILDER_DEPOSIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007732` <!-- placeholder --> | Predeploy address of the builder deposit contract |
| `BUILDER_EXIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007733` <!-- placeholder --> | Predeploy address of the builder exit contract |
| `BUILDER_DEPOSIT_REQUEST_TYPE` | `0x03`  <!-- placeholder --> | [EIP-7685](./eip-7685.md) request-type byte for builder deposits |
| `BUILDER_EXIT_REQUEST_TYPE` | `0x04` <!-- placeholder --> | [EIP-7685](./eip-7685.md) request-type byte for builder exits  |
```

| `TARGET_DEPOSIT_REQUESTS_PER_BLOCK` | `32` | Per-block request count above which the fee rises for the deposit contract |
| `MAX_EXIT_REQUESTS_PER_BLOCK` | `16` | Maximum records the builder exit contract drains into one block |
| `TARGET_EXIT_REQUESTS_PER_BLOCK` | `2` | Per-block request count above which the fee rises for the exit contract |
| `MIN_REQUEST_FEE` | `1` | Minimum request fee, in wei |
| `REQUEST_FEE_UPDATE_FRACTION` | `17` | Controls the fee's rate of change |
| `EXCESS_INHIBITOR` | `2**256-1` | Excess value that makes the fee getter revert before the first system call (as in [EIP-7002](./eip-7002.md)/[EIP-7251](./eip-7251.md)); set at deployment, cleared by the first system call |
| `BUILDER_MIN_DEPOSIT` | `1000000000000000000` | Minimum credited stake for a deposit, in wei (1 ETH — the [EIP-7732](./eip-7732.md) builder minimum) |
| `BUILDER_DEPOSIT_CONTRACT_RUNTIME_CODE` | *see [Reference Implementation](#reference-implementation)* | Runtime bytecode of the builder deposit contract |
| `BUILDER_EXIT_CONTRACT_RUNTIME_CODE` | *see [Reference Implementation](#reference-implementation)* | Runtime bytecode of the builder exit contract |

### Deployment

> Each predeploy is deployed exactly as the [EIP-7002](./eip-7002.md) and [EIP-7251](./eip-7251.md) request contracts are: by a one-time presigned transaction from a single-use deployer account (the Nick's-method scheme), so that `BUILDER_DEPOSIT_CONTRACT_ADDRESS` and `BUILDER_EXIT_CONTRACT_ADDRESS` are the addresses cryptographically derived from those transactions. Each contract's init code sets its `excess` slot to `EXCESS_INHIBITOR`, so no request can be enqueued until the inhibitor is cleared (see [Request fee](#request-fee)). The concrete transactions — and therefore the final addresses — will be fixed once the runtime bytecode is audited and frozen (see [Reference Implementation](#reference-implementation)).

Reviewer 3: > (the Nick's-method scheme)

If there's somewhere we can link to that defines this in detail (preferably in another EIP/ERC), we should.

This EIP is additive at the execution layer: it introduces new contracts at previously empty addresses. It does not modify the validator deposit contract at `0x00000000219ab540356cbb839cbe05303d7705fa`, the validator withdrawal/consolidation predeploys, or any existing validator's lifecycle.

At the consensus layer it modifies EIP-7732 (see [Changes to EIP-7732](#changes-to-eip-7732)): post-fork builder onboarding moves from the validator deposit request to `BUILDER_DEPOSIT_REQUEST_TYPE`, and builder exits move from the voluntary-exit operation to `BUILDER_EXIT_REQUEST_TYPE`. The fork-transition onboarding of builder-credentialed pending deposits is unchanged, so builders present at the fork are unaffected. The new request types are additive — blocks that contain no builder requests produce empty `request_data` for these types, which [EIP-7685](./eip-7685.md) excludes from the `requests_hash`.

## Test Cases

Test cases will be provided with the reference implementation, once the predeploy contracts are finalized.

## Reference Implementation

The contracts will be developed and finalized as system-contract sources, against which the deployed runtime bytecode must reproduce.

> The final `BUILDER_DEPOSIT_CONTRACT_RUNTIME_CODE` and `BUILDER_EXIT_CONTRACT_RUNTIME_CODE`, the predeploy addresses, and the request-type bytes will be locked in once the contracts have been independently audited. The runtime bytecode, the exact compiler version and settings used to produce it, and the contracts' storage layout MUST be pinned together at that point, so the canonical bytecode is independently reproducible.

Reviewer 3: ```suggestion
<!-- TODO: Add test cases and reference implementation -->
```

Again, write as if everything is already complete.

- **Exit authorization.** The exit contract records `msg.sender` as `source_address` and performs no further check. Because the request carries no signature, this is the sole authorization: the consensus layer MUST initiate an exit only when `source_address` equals the target builder's `execution_address`, or an arbitrary caller could exit a builder it does not control. A builder's only exit authorizer is therefore its `execution_address`; the voluntary-exit (BLS-key) path is removed for builders.

> - **Custodial-split exit standoff.** A builder's exit precondition requires its pending balance to be zero (`get_pending_balance_to_withdraw_for_builder == 0`), every winning bid adds a pending payment, and the `execution_address` is the builder's sole exit authorizer (the BLS voluntary-exit path is removed). When the `execution_address` (the capital owner) and the BLS key (the bidding operator) are held by different parties — a custodial or staking-pool arrangement this design explicitly enables — the operator can keep the pending balance non-zero by continuing to win bids, so the capital owner cannot satisfy the exit precondition and the stake stays locked (a builder that never exits is never swept). The standoff is self-limiting, since the operator's bids must keep being included on-chain, but the protocol gives the `execution_address` holder no on-chain lever to halt bidding. Parties delegating builder operation SHOULD retain off-chain (contractual or operational) control over the operator's bidding, so a delegated builder can always be brought to a state in which it can exit.

Reviewer 3: Avoid creating requirements (defined with UPPERCASE keywords) outside of the Specification section. They're too easy to miss buried in this section. Instead, move the requirement itself to the Spec section, and discuss its security implications here.

## Motivation

[EIP-7732](./eip-7732.md) introduces builders as a separate, staked consensus-layer class. A builder is created by a deposit, can have stake added, and must be able to exit. Today EIP-7732 sources this lifecycle from the *validator* flows: a builder is registered by an ordinary validator deposit request whose withdrawal credential carries the `0x03` `BUILDER_WITHDRAWAL_PREFIX`, and a builder exits through a builder branch of the consensus-layer voluntary-exit operation. This EIP instead gives builders their own dedicated [EIP-7685](./eip-7685.md) request contracts.

> **Dedicated request types remove cross-actor coupling.** Routing builders through the validator contracts forces the consensus layer to decide, on every request, whether it acts on the validator set or the builder set (today by inspecting the credential prefix). Dedicated builder request types make the actor explicit from the request type alone, so the validator and builder registries are keyed independently. A single public key can then be registered as both a validator and a builder; the protocol currently disallows that overlap, and this EIP allows the rule to be removed.

Reviewer 4: > A single public key can then be registered as both a validator and a builder; the protocol currently disallows that overlap, and this EIP allows the rule to be removed.

Do you think if it's worth to mention whether this removal has any concern?

> **The deposit bounds a consensus-side denial-of-service surface.** A builder deposit's proof-of-possession is verified by the consensus layer (as it already is in EIP-7732). Routed through the validator deposit request — which admits thousands of deposits per block — an attacker submitting invalid-signature builder deposits at the 1-ETH builder minimum could force that many proof-of-possession checks per block. Delivering builder deposits through a *dedicated* request bus caps them at `MAX_REQUESTS_PER_BLOCK` per block and charges an [EIP-1559](./eip-1559.md)-style fee on top of the staked value, bounding both the consensus-layer verification work and the spam economics.

Reviewer 4: This might be interpreted as capping max builder deposits per block solves the signature verification burden. Before EIP-8282, the burden comes from the work of verifying signatures of all pending validators. The capping doesn't solve this issue technically. What solves it is a separation of validator and builder queues, and capping here bounds the builder's signatures verification time.

### Constants

All address and request-type values below are placeholders. The `0x03`/`0x04` request types MUST be unique across **all** active [EIP-7685](./eip-7685.md) request types, including in-flight proposals ([EIP-7804](./eip-7804.md), a Draft, also claims `0x03`); final allocation is coordinated in consensus-specs.

| Name | Value | Comment |
| --- | --- | --- |
| `BUILDER_DEPOSIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007732` | Predeploy address of the builder deposit contract (placeholder) |
| `BUILDER_EXIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007733` | Predeploy address of the builder exit contract (placeholder) |
| `BUILDER_DEPOSIT_REQUEST_TYPE` | `0x03` | [EIP-7685](./eip-7685.md) request-type byte for builder deposits (placeholder) |
| `BUILDER_EXIT_REQUEST_TYPE` | `0x04` | [EIP-7685](./eip-7685.md) request-type byte for builder exits (placeholder) |
| `SYSTEM_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe` | Address that invokes the end-of-block system call (as in [EIP-7002](./eip-7002.md)) |
| `MAX_REQUESTS_PER_BLOCK` | `16` | Maximum records each contract drains into one block |

> | `TARGET_REQUESTS_PER_BLOCK` | `2` | Per-block request count above which the fee rises |

Reviewer 5: ```suggestion
| `MAX_DEPOSIT_REQUESTS_PER_BLOCK` | `256` | Maximum records the builder deposit contract drains into one block |
| `TARGET_DEPOSIT_REQUESTS_PER_BLOCK` | `32` | Per-block request count above which the fee rises for the deposit contract |
| `MAX_EXIT_REQUESTS_PER_BLOCK` | `16` | Maximum records the builder exit contract drains into one block |
| `TARGET_EXIT_REQUESTS_PER_BLOCK` | `2` | Per-block request count above which the fee rises for the exit contract |
```
To match https://github.com/ethereum/sys-asm/pull/43/changes/537b9c17d14705e77bc3f512e17dd253ad69b020

    signature: Bytes96

class BuilderExitRequest(object):
    source_address: Bytes20
    pubkey: Bytes48
```

A type's `request_data` is the concatenation of the fixed-size SSZ serializations of its records — 184 bytes per `BuilderDepositRequest` (`pubkey ++ withdrawal_credentials ++ amount ++ signature`) and 68 bytes per `BuilderExitRequest` (`source_address ++ pubkey`), with `amount` little-endian — exactly the bytes the system call returns, in the same order. `BuilderDepositRequest` is the validator [EIP-6110](./eip-6110.md) `DepositRequest` without the `index` field; the consensus layer verifies its `signature` (the proof-of-possession) on the builder's first registration.

### Consensus-layer processing of records

Both request types are applied immediately when processed — a `BuilderDepositRequest` is **not** routed through the validator `pending_deposits` queue, so a builder's balance is credited without an activation-churn queue, preserving EIP-7732's existing behavior. (A newly registered builder still becomes active for bidding and exit only once its deposit epoch is finalized, per gloas `is_active_builder`; only the churn queue is skipped, not finality.)

> - A `BuilderDepositRequest` (type `0x03`) for a `pubkey` **not** yet in the builder set is a first deposit. The consensus layer registers the builder only if both checks pass: its `withdrawal_credentials` begins with the `0x03` `BUILDER_WITHDRAWAL_PREFIX` (`is_builder_withdrawal_credential`), and the proof-of-possession `signature` over the `DepositMessage` `(pubkey, withdrawal_credentials, amount)` under `DOMAIN_DEPOSIT` is valid — the same signature check validator deposits use (gloas `is_valid_deposit_signature`). If both hold, it registers the builder with the record's `withdrawal_credentials` (whose last 20 bytes are the builder's `execution_address`) and credits its `amount`. A record whose `withdrawal_credentials` is not `0x03`-prefixed, or whose signature is invalid, is ignored (consumed, stake forfeited). The prefix check mirrors the credential check `process_deposit_request` applies on the validator path, so a registered builder always carries a `0x03` credential and therefore a well-formed `execution_address`.

Reviewer 4: ```suggestion
- A `BuilderDepositRequest` (type `0x03`) for a `pubkey` **not** yet in the builder set is a first deposit. The consensus layer registers the builder only if both checks pass: its `withdrawal_credentials` begins with the `0x03` `BUILDER_WITHDRAWAL_PREFIX` (`is_builder_withdrawal_credential`), and the proof-of-possession `signature` over the `DepositMessage` `(pubkey, withdrawal_credentials, amount)` under `DOMAIN_BUILDER_DEPOSIT` is valid — the same signature check validator deposits use (gloas `is_valid_deposit_signature`). If both hold, it registers the builder with the record's `withdrawal_credentials` (whose last 20 bytes are the builder's `execution_address`) and credits its `amount`. A record whose `withdrawal_credentials` is not `0x03`-prefixed, or whose signature is invalid, is ignored (consumed, stake forfeited). The prefix check mirrors the credential check `process_deposit_request` applies on the validator path, so a registered builder always carries a `0x03` credential and therefore a well-formed `execution_address`.
```
According to the current spec PR, it uses `DOMAIN_DEPOSIT` for onboarding at fork and `DOMAIN_BUILDER_DEPOSIT` otherwise.

Reviewer 6: A few nits:

> "(gloas `is_valid_deposit_signature`)"

This is not a function introduced in Gloas. I'd just drop the "gloas" word.

> A record whose `withdrawal_credentials` is not `0x03`-prefixed, or whose signature is invalid, is ignored (consumed, stake forfeited).

Not true. Any prefix (version) is valid. They could use 0x00 - 0xFF.

> The prefix check mirrors the credential check `process_deposit_request` applies on the validator path, so a registered builder always carries a `0x03` credential and therefore a well-formed `execution_address`.

I would just delete this.

> - A `BuilderDepositRequest` (type `0x03`) for a `pubkey` **not** yet in the builder set is a first deposit. The consensus layer registers the builder only if both checks pass: its `withdrawal_credentials` begins with the `0x03` `BUILDER_WITHDRAWAL_PREFIX` (`is_builder_withdrawal_credential`), and the proof-of-possession `signature` over the `DepositMessage` `(pubkey, withdrawal_credentials, amount)` under `DOMAIN_DEPOSIT` is valid — the same signature check validator deposits use (gloas `is_valid_deposit_signature`). If both hold, it registers the builder with the record's `withdrawal_credentials` (whose last 20 bytes are the builder's `execution_address`) and credits its `amount`. A record whose `withdrawal_credentials` is not `0x03`-prefixed, or whose signature is invalid, is ignored (consumed, stake forfeited). The prefix check mirrors the credential check `process_deposit_request` applies on the validator path, so a registered builder always carries a `0x03` credential and therefore a well-formed `execution_address`.

Reviewer 4: > its `withdrawal_credentials` begins with the `0x03` `BUILDER_WITHDRAWAL_PREFIX` (`is_builder_withdrawal_credential`)

This holds only for onboarding builders at fork.

> The prefix check mirrors the credential check `process_deposit_request` applies on the validator path

This is a bit misleading because `process_deposit_request` doesn't apply any credential check. It stores a credential as-is and the credential is checked during withdrawal.

Reviewer 4: Additionally, I've found withdrawal_credentials[0] is used as a builder version afterward. I think it's good to mention this somewhere. Maybe it can piggyback on this EIP? Not entirely sure though.

- A `BuilderDepositRequest` (type `0x03`) for a `pubkey` **already** in the builder set is a top-up: it credits `amount`, and its `withdrawal_credentials` and `signature` MUST be ignored — the existing registration is unchanged. This mirrors the validator deposit contract, where the proof-of-possession is checked only on a pubkey's first appearance and later deposits are stake additions. The builder set still contains entries that have **exited** (a slot is reclaimed only once the builder's `withdrawable_epoch` has passed and its balance is zero), so a deposit to an exited `pubkey` is also a top-up — it credits an entry that EIP-7732 does not reactivate, so the added stake sweeps to that entry's `execution_address` and never resumes bidding; re-registering the key requires waiting for its prior slot to be recycled.

> - A `BuilderExitRequest` (type `0x04`) MUST be ignored unless its `pubkey` is a registered, active builder (gloas `is_active_builder`: its deposit epoch is finalized and it is not already exiting), its `source_address` equals that builder's `execution_address`, and it has no pending balance to withdraw (`get_pending_balance_to_withdraw_for_builder == 0`). This is precisely EIP-7732's `process_voluntary_exit` builder branch with the BLS-signature check replaced by the `source_address` check; when the predicate holds it runs `initiate_builder_exit` (`withdrawable_epoch = current_epoch + MIN_BUILDER_WITHDRAWABILITY_DELAY`). A request that fails any precondition is **consumed and discarded, not re-queued** — the fee is spent. Because an active builder routinely has a non-zero pending balance from recent bid payments, a legitimate exit may be dropped until those settle, in which case the caller must resubmit once the pending balance has been swept. (The execution layer dequeues the record deterministically regardless, so a dropped request never affects `requests_hash` agreement.)

Reviewer 4: I think the builder exit is analogous to `process_withdrawal_request`, not `process_voluntary_exit`. It's triggered by execution request. So it needs to check whether the credential is 0x03. cc @jtraglia

Reviewer 6: Hmm not sure I agree. I think it is more like a voluntary exit. The only reason we'd want to check that the version is a specific value would be to incentivize builders use supported versions; ie don't make a builder with version=0xFF. But I don't think this is a strong enough reason to do this, as I expect some builders to accidentally use the 0x03 withdrawal credential with the new contract.

### Changes to EIP-7732

This EIP modifies EIP-7732's builder lifecycle on the consensus layer:

> - **Deposit routing.** `process_deposit_request` no longer creates or tops up builders. Its builder branch — `if is_builder or (is_builder_withdrawal_credential(...) and not is_validator and not is_pending_validator)` → `apply_deposit_for_builder` — is replaced by `if is_builder_withdrawal_credential(deposit_request.withdrawal_credentials): return`: the deposit is inert — it is **not** appended to `pending_deposits` (so no validator is minted) and its ETH is forfeited in the immutable deposit contract, as with any misdirected deposit. All other deposits process as validator deposits unchanged. Consequently a validator-contract deposit to a `pubkey` that is already a builder now creates or credits a **validator** with that key (the same key may be both — see [Rationale](#rationale)), never the builder; builders are created and topped up **only** via `BUILDER_DEPOSIT_REQUEST_TYPE`.

Reviewer 4: > is replaced by `if is_builder_withdrawal_credential(deposit_request.withdrawal_credentials): return`: the deposit is inert — it is **not** appended to `pending_deposits` (so no validator is minted) and its ETH is forfeited in the immutable deposit contract, as with any misdirected deposit.

This is not what the spec PR does. It doesn't change `process_deposit_request` at all, i.e., it doesn't check whether it's a builder deposit. It will create a new validator but it will be impossible to withdraw. My understanding is `process_deposit_request` assumes the request is well-formed. cc @jtraglia

Reviewer 4: > Consequently a validator-contract deposit to a `pubkey` that is already a builder now creates or credits a **validator** with that key (the same key may be both — see [Rationale](#rationale))

Which bullet in the rationale section does this refer to?

Reviewer 6: Yeah part in your first comment is 100% incorrect.

> - **Genesis onboarding (at the fork).** `onboard_builders_from_pending_deposits`, run once during the fork upgrade, is retained: builder-credentialed deposits already in `pending_deposits` at the upgrade are onboarded as builders, so builders exist from the activation slot. Operators seed the genesis set by depositing to the existing deposit contract with a `BUILDER_WITHDRAWAL_PREFIX` credential before the fork — late enough that the deposit is still pending at the upgrade (a deposit applied earlier would create a stranded validator). The cutover is a single deterministic switch: deposits captured by the snapshot are onboarded, and from the fork onward the deposit-routing rule above drops every `0x03`-credentialed validator-contract deposit. A `0x03`-credentialed deposit that lands too late for the snapshot is therefore **not** onboarded; its stake is forfeited like any other dropped deposit, and the operator re-onboards through `BUILDER_DEPOSIT_REQUEST_TYPE`. No `pubkey` is onboarded by more than one path.

Reviewer 4: > drops every `0x03`-credentialed validator-contract deposit

I think it's not dropped. It's a valid but malformed validator deposit.

Reviewer 2: pretty sure as the spec works right now it would just be credited to the validator anyways?

- **Exit by `execution_address`; voluntary exit becomes validator-only.** A builder's BLS key is hot — it signs bids continuously — so authorizing exit with that key is undesirable. Routing exit through the `execution_address` (the cold address that owns the builder's stake and receives its withdrawals) mirrors EIP-7002's rationale for letting `0x01` credentials trigger validator exits, and removing the builder branch from the voluntary-exit operation gives builders a single, well-defined exit authorizer. Losing the `execution_address` key strands no funds that were not already stranded: that address is where the builder's balance is swept regardless.

- **EIP-1559-style request fee.** Each request carries the same demand-responsive fee as EIP-7002/7251: super-linear above `TARGET_REQUESTS_PER_BLOCK`, decaying back to `MIN_REQUEST_FEE` when demand subsides. Together with the per-block cap and the per-deposit stake, the fee meters submission to each predeploy.

> - **Genesis onboarding via the fork transition.** Some applications depend on builders existing from the first slot of the fork. EIP-7732 already onboards builder-credentialed pending deposits during the fork upgrade; retaining that — rather than relying on post-fork deposits to the new contract, which cannot populate the activation slot — keeps the genesis builder set available immediately. That one-time onboarding is bounded, so it needs neither the per-block cap nor the fee that the steady-state contract provides.

Reviewer 4: Just a nit, "genesis" here might be not a super precise term.

Reviewer 4: > That one-time onboarding is bounded, so it needs neither the per-block cap nor the fee that the steady-state contract provides.

How is the one-time onboarding is bounded?

Reviewer 6: Good to call out. As long as the pending deposit queue is sufficiently large, it is effectively unbounded.

### Constants

All address and request-type values below are placeholders. The `0x03`/`0x04` request types MUST be unique across **all** active [EIP-7685](./eip-7685.md) request types, including in-flight proposals ([EIP-7804](./eip-7804.md), a Draft, also claims `0x03`); final allocation is coordinated in consensus-specs.

| Name | Value | Comment |
| --- | --- | --- |
| `BUILDER_DEPOSIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007732` | Predeploy address of the builder deposit contract (placeholder) |
| `BUILDER_EXIT_CONTRACT_ADDRESS` | `0x0000000000000000000000000000000000007733` | Predeploy address of the builder exit contract (placeholder) |
| `BUILDER_DEPOSIT_REQUEST_TYPE` | `0x03` | [EIP-7685](./eip-7685.md) request-type byte for builder deposits (placeholder) |
| `BUILDER_EXIT_REQUEST_TYPE` | `0x04` | [EIP-7685](./eip-7685.md) request-type byte for builder exits (placeholder) |
| `SYSTEM_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe` | Address that invokes the end-of-block system call (as in [EIP-7002](./eip-7002.md)) |
| `MAX_DEPOSIT_REQUESTS_PER_BLOCK` | `256` | Maximum records the builder deposit contract drains into one block |

> | `TARGET_DEPOSIT_REQUESTS_PER_BLOCK` | `32` | Per-block request count above which the fee rises for the deposit contract |

Reviewer 2: we might still wanna tweak these, see https://github.com/ethereum/sys-asm/pull/43#discussion_r3470524501

- **Exit routing.** gloas no longer overrides `process_voluntary_exit` — its former builder branch is removed, making the voluntary-exit operation validator-only again — and builders exit only via `BUILDER_EXIT_REQUEST_TYPE`, handled by the new `process_builder_exit_request`.

## Rationale

> - **Copied from the audited request contracts.** Both predeploys reuse the source of the deployed [EIP-7002](./eip-7002.md) and [EIP-7251](./eip-7251.md) system contracts (the `geas` implementations maintained in the `ethereum/sys-asm` repository, which share one code skeleton) rather than introducing new contract code. The dispatch, fee mechanism, queue, storage layout, and system subroutine carry over line-for-line — and with them the raw fixed-size calldata interface (no ABI) and the per-request log. The exit contract is the EIP-7002 withdrawals contract with the amount field removed; the deposit contract is the same skeleton with a wider record, no source-address prefix, and two added value checks, reusing EIP-7002's big-endian-in/little-endian-out amount convention and its conversion code. These deltas — request layout constants plus a handful of comparison instructions — are the entire audit burden the two new contracts add.

Reviewer 3: > the `geas` implementations maintained in the `ethereum/sys-asm` repository, which share one code skeleton

This is definitely an external link. I'd remove it if you can, or apply following the EIP-5757 process to get it allowed.

- **Exit by `execution_address`; voluntary exit becomes validator-only.** A builder's BLS key is hot — it signs bids continuously — so authorizing exit with that key is undesirable. Routing exit through the `execution_address` (the cold address that owns the builder's stake and receives its withdrawals) mirrors EIP-7002's rationale for letting `0x01` credentials trigger validator exits, and removing the builder branch from the voluntary-exit operation gives builders a single, well-defined exit authorizer. Losing the `execution_address` key strands no funds that were not already stranded: that address is where the builder's balance is swept regardless.

> - **EIP-1559-style request fee.** Each request carries the same demand-responsive fee as EIP-7002/7251: super-linear above `TARGET_REQUESTS_PER_BLOCK`, decaying back to `MIN_REQUEST_FEE` when demand subsides. Together with the per-block cap and the per-deposit stake, the fee meters submission to each predeploy.

Reviewer 3: ```suggestion
- **EIP-1559-style request fee.** Each request carries the same demand-responsive fee as EIP-7002/EIP-7251: super-linear above `TARGET_REQUESTS_PER_BLOCK`, decaying back to `MIN_REQUEST_FEE` when demand subsides. Together with the per-block cap and the per-deposit stake, the fee meters submission to each predeploy.
```

> ---

Reviewer 7: ```suggestion
---
eip: 8282
```
@wemeetagain 

Assigning next sequential EIP/ERC/RIP number.
Numbers are assigned by editors & associates.

Please also update the filename.
